File: java/diffie-hellman/src/main/java/DiffieHellman.java
import java.math.BigInteger;
import java.util.Random;

class DiffieHellman {

    BigInteger privateKey(BigInteger prime) {
        // random number between 1 (inclusive) and prime (exclusive)
        int rand = 1 + new Random().nextInt(prime.intValue() - 1);
        return BigInteger.valueOf(rand);
    }

    BigInteger publicKey(BigInteger p, BigInteger g, BigInteger privateKey) {
        return g.modPow(privateKey, p);
    }

    BigInteger secret(BigInteger p, BigInteger publicKey, BigInteger privateKey) {
        return publicKey.modPow(privateKey, p);
    }
}

File: kotlin/diffie-hellman/src/main/kotlin/DiffieHellman.kt
import java.math.BigInteger
import kotlin.random.Random

object DiffieHellman {

    fun privateKey(prime: BigInteger): BigInteger =
        BigInteger.valueOf(1L + Random.nextLong(prime.longValueExact() - 1))

    fun publicKey(p: BigInteger, g: BigInteger, privateKey: BigInteger): BigInteger =
        g.modPow(privateKey, p)

    fun secret(p: BigInteger, publicKey: BigInteger, privateKey: BigInteger): BigInteger =
        publicKey.modPow(privateKey, p)
}
